Utils/src/BTaggingEffAnalyzer.cc

```diff
@@ -0,0 +1,233 @@
+  ptNBins(iConfig.getParameter<int>("PtNBins")),
+  ptMin(iConfig.getParameter<double>("PtMin")),
+  ptMax(iConfig.getParameter<double>("PtMax")),
+  etaNBins(iConfig.getParameter<int>("EtaNBins")),
+  etaMin(iConfig.getParameter<double>("EtaMin")),
+  etaMax(iConfig.getParameter<double>("EtaMax"))
+
+{
+   //now do what ever initialization is needed
+   h2_BTaggingEff_Denom_b    = fs->make<TH2D>("h2_BTaggingEff_Denom_b", ";p_{T} [GeV];#eta", ptNBins, ptMin, ptMax, etaNBins, etaMin, etaMax);
+   h2_BTaggingEff_Denom_c    = fs->make<TH2D>("h2_BTaggingEff_Denom_c", ";p_{T} [GeV];#eta", ptNBins, ptMin, ptMax, etaNBins, etaMin, etaMax);
+   h2_BTaggingEff_Denom_udsg = fs->make<TH2D>("h2_BTaggingEff_Denom_udsg", ";p_{T} [GeV];#eta", ptNBins, ptMin, ptMax, etaNBins, etaMin, etaMax);
+   h2_BTaggingEff_Num_b    = fs->make<TH2D>("h2_BTaggingEff_Num_b", ";p_{T} [GeV];#eta", ptNBins, ptMin, ptMax, etaNBins, etaMin, etaMax);
+   h2_BTaggingEff_Num_c    = fs->make<TH2D>("h2_BTaggingEff_Num_c", ";p_{T} [GeV];#eta", ptNBins, ptMin, ptMax, etaNBins, etaMin, etaMax);
+   h2_BTaggingEff_Num_udsg = fs->make<TH2D>("h2_BTaggingEff_Num_udsg", ";p_{T} [GeV];#eta", ptNBins, ptMin, ptMax, etaNBins, etaMin, etaMax);
+}
+
+BTaggingEffAnalyzer::~BTaggingEffAnalyzer()
+{
+ 
+   // do anything here that needs to be done at desctruction time
+   // (e.g. close files, deallocate resources etc.)
+
+}
+
+
+//
+// member functions
+//
+
+// ------------ method called for each event  ------------
+void
+BTaggingEffAnalyzer::analyze(const edm::Event& iEvent, const edm::EventSetup& iSetup)
+{
+  edm::Handle<PatJetCollection> jets;
+  iEvent.getByLabel(jetsTag,jets);
+
+  //grab event selection variables:
+  edm::Handle<double> HT;
+  edm::Handle<double> MHT; 
+  edm::Handle<double>DeltaPhi1;
+  edm::Handle<double>DeltaPhi2;
+  edm::Handle<double>DeltaPhi3;
+  edm::Handle<double>DeltaPhi4;
+  edm::Handle<int> NJets;
+  edm::Handle<int> Leptons;
+  edm::Handle<int> isoPionTracks;
+  edm::Handle<int> isoMuonTracks;
+  edm::Handle<int> isoElectronTracks;
+  edm::Handle<bool> JetID; 
+
+  iEvent.getByLabel("HT", HT);
+  iEvent.getByLabel("MHT","Pt", MHT);
+  iEvent.getByLabel("DeltaPhi", "DeltaPhi1", DeltaPhi1);
+  iEvent.getByLabel("DeltaPhi","DeltaPhi2", DeltaPhi2);
+  iEvent.getByLabel("DeltaPhi","DeltaPhi3", DeltaPhi3);
+//  iEvent.getByLabel("DeltaPhi","DeltaPhi4", DeltaPhi4);
+
+  iEvent.getByLabel("NJets", NJets);
+  iEvent.getByLabel("LeptonsNew", Leptons);
+  iEvent.getByLabel("IsolatedPionTracksVeto","isoTracks", isoPionTracks);
+  iEvent.getByLabel("IsolatedMuonTracksVeto","isoTracks", isoMuonTracks);
+  iEvent.getByLabel("IsolatedElectronTracksVeto","isoTracks", isoElectronTracks);
+  iEvent.getByLabel("GoodJets", JetID);
+  bool passEvSel=false;
+  if(*HT>500 && *MHT>200 && *DeltaPhi1>0.5 && *DeltaPhi2>0.5 && *DeltaPhi3>0.3 && *NJets>=4 && *Leptons==0 && *isoPionTracks==0 && *isoMuonTracks==0 && *isoElectronTracks==0 && *JetID)passEvSel=true;
```

Comment:
My concern is that, conceptually, the goal of TreeMaker is flattening ntuples, applying object IDs, and calculating related quantities from identified objects. Applying event selections and running analyzers is intended to happen downstream. Since we already store the b-tag discriminator value and parton flavor for all the GoodJets, I think this entire procedure fits better into user analysis code outside of TreeMaker... for example, if you use my signal skims, the duplication of the signal selection can be entirely avoided, and you just have to apply a few cuts on the jet collections. I could even write this into my analysis code pretty easily, to keep things organized.